org.alloytools.alloy.cli/src/main/java/org/alloytools/alloy/cli/CLI.java

```diff
@@ -0,0 +1,259 @@
+		DecomposeStrategy decompose_mode(DecomposeStrategy dflt);
+
+		/**
+		 * This option specifies the number of threads when following a
+		 * decompose strategy
+		 * <p>
+		 * Default value is 4.
+		 */
+		@Description("This option specifies the number of threads when following a decompose strategy. Default value is 4.")
+		int decompose_threads(int deflt);
+
+		/**
+		 * Set trace level
+		 */
+
+		boolean trace();
+
+		/**
+		 * Set exceptions
+		 */
+
+		boolean exceptions();
+	}
+
+	public void __run(CLIOptions options) throws Exception {
+
+		if (options.trace()) {
+			setTrace(true);
+		}
+		if (options.exceptions()) {
+			setExceptions(true);
+		}
+		this.options.coreMinimization = options
+				.coreMinimization(CoreMininmization.values()[this.options.coreMinimization]).ordinal();
+		this.options.decompose_mode = options.decompose_mode(DecomposeStrategy.values()[this.options.decompose_mode])
+				.ordinal();
+		this.options.decompose_threads = options.decompose_threads(this.options.decompose_threads);
+		this.options.inferPartialInstance = options.inferPartialInstance(this.options.inferPartialInstance);
+		this.options.noOverflow = options.noOverflow(this.options.noOverflow);
+		this.options.skolemDepth = options.skolemDepth(this.options.skolemDepth);
+		String solver = options.solver(this.options.solver.id());
+		this.options.solver = SatSolver.parse(solver);
+		if (!this.options.solver.id().equals(solver)) {
+			warning("The solver name %s not found, using %s. Use `alloy solvers` to see a list of all solvers",
+					solver, this.options.solver);
+		}
+		this.options.symmetry = options.breakingSymmetry(this.options.symmetry);
+		this.options.unrolls = options.recurse(this.options.unrolls);
+
+		List<String> args = options._arguments();
+		if (args.isEmpty()) {
+			Justif j = new Justif();
+			cl.help(j.formatter(), this);
+			System.out.println(j.wrap());
+			return;
+		}
+
+		String help = cl.execute(this, args.remove(0), args);
+		if (help != null) {
+			System.out.println(help);
+		}
+		report(System.out);
+
+	}
+
+	/**
+	 * Show the list of solvers
+	 */
+
+	@Arguments(arg = {})
+	@Description("Show the list of solvers")
+	interface SolverOptions extends Options {
+		@Description("Show extended information")
+		boolean xtended();
+	}
+
+	public void _solvers(SolverOptions options) {
+		SafeList<SatSolver> values = SatSolver.values();
+		values.forEach(s -> {
+			if (options.xtended()) {
+				System.out.printf("%-20s %-20s %s\n", s.id(), maskNull(s.external()), Arrays.toString(s.options()));
+			} else {
+				System.out.println(s.id());
+			}
+		});
+
+	}
+
+	private String maskNull(Object s) {
+		if (s == null) {
+			return "";
+		}
+		return s.toString();
+	}
+
+	/**
+	 * Run all 'run' & asserts
+	 */
+
+	interface ExecOptions extends Options {
+
+	}
+
+	public void _exec(ExecOptions options) throws IOException {
+		SimpleReporter rep = new SimpleReporter(this);
+		String filename = options._arguments().remove(0);
```

Comment:
I realize this is a draft PR, so you may already be aware of this and intending to fix this so feel free to ignore this comment. At the moment, this line causes an unhandled exception if the user fails to specify a file:
```
$ java -jar org.alloytools.alloy.cli/target/org.alloytools.alloy.cli.jar exec
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:359)
	at java.base/java.util.ArrayList.remove(ArrayList.java:504)
	at org.alloytools.alloy.cli.CLI._exec(CLI.java:221)
	at aQute.lib.getopt.CommandLine.execute(CommandLine.java:158)
	at org.alloytools.alloy.cli.CLI.__run(CLI.java:173)
	at aQute.lib.getopt.CommandLine.execute(CommandLine.java:158)
	at org.alloytools.alloy.cli.CLI.main(CLI.java:257)
```
I think it would be more informative/user-friendly to print a message saying a file is required.